src/OpenTelemetry.Instrumentation.EventCounters/EventCountersMetrics.cs

```diff
@@ -0,0 +1,137 @@
+
+    private readonly EventCountersInstrumentationOptions options;
+    private readonly ConcurrentQueue<EventSource> preInitEventSources = new();
+    private readonly ConcurrentDictionary<(string, string), Instrument> instruments = new();
+    private readonly ConcurrentDictionary<(string, string), double> values = new();
+
+    /// <summary>
+    /// Initializes a new instance of the <see cref="EventCountersMetrics"/> class.
+    /// </summary>
+    /// <param name="options">The options to define the metrics.</param>
+    public EventCountersMetrics(EventCountersInstrumentationOptions options)
+    {
+        this.options = options;
+
+        while (this.preInitEventSources.TryDequeue(out EventSource eventSource))
+        {
+            if (this.options.ShouldListenToSource(eventSource.Name))
+            {
+                this.EnableEvents(eventSource, EventLevel.LogAlways, EventKeywords.None, GetEnableEventsArguments(this.options));
+            }
+        }
+    }
+
+    /// <inheritdoc />
+    protected override void OnEventSourceCreated(EventSource source)
+    {
+        if (this.options == null)
+        {
+            this.preInitEventSources.Enqueue(source);
+        }
+        else if (this.options.ShouldListenToSource(source.Name))
+        {
+            this.EnableEvents(source, EventLevel.LogAlways, EventKeywords.None, GetEnableEventsArguments(this.options));
+        }
+    }
+
+    /// <inheritdoc />
+    protected override void OnEventWritten(EventWrittenEventArgs eventData)
+    {
+        if (this.options == null)
+        {
+            return;
+        }
+
+        var eventSourceName = eventData.EventSource.Name;
+
+        if (eventData.EventName != "EventCounters")
+        {
+            EventCountersInstrumentationEventSource.Log.IgnoreNonEventCountersName(eventSourceName);
+            return;
+        }
+
+        if (eventData.Payload == null || eventData.Payload.Count == 0 || eventData.Payload[0] is not IDictionary<string, object> payload)
+        {
+            EventCountersInstrumentationEventSource.Log.IgnoreEventWrittenEventArgsPayloadNotParseable(eventSourceName);
+            return;
+        }
+
+        var hasName = payload.TryGetValue("Name", out var nameObj);
+
+        if (!hasName)
+        {
+            EventCountersInstrumentationEventSource.Log.IgnoreEventWrittenEventArgsWithoutName(eventSourceName);
+            return;
+        }
+
+        var name = nameObj.ToString();
+
+        var hasMean = payload.TryGetValue("Mean", out var mean);
+        var hasIncrement = payload.TryGetValue("Increment", out var increment);
+
+        if (!(hasIncrement ^ hasMean))
+        {
+            EventCountersInstrumentationEventSource.Log.IgnoreMeanIncrementConflict(eventSourceName);
+            return;
+        }
+
+        var value = Convert.ToDouble(hasMean ? mean : increment);
```

Comment:
You could move the `try-catch` block from `UpdateInstrumentWithEvent` to here to include the `Convert.ToDouble` method in the `try` block as that could also throw exceptions.